Task 2. Regex to DFA

DavidBykov opened the pull request:
- [X] Используя возможности [pyformlang](https://pyformlang.readthedocs.io/en/latest/) реализовать **функцию** построения минимального ДКА по заданному регулярному выражению. [Формат регулярного выражения.](https://pyformlang.readthedocs.io/en/latest/usage.html#regular-expression).
- [X] Добавить необходимые тесты.

DavidBykov committed:
- task_2

EgorOrachyov commented on project/regex.py:
```diff
@@ -0,0 +1,24 @@
+from pyformlang.finite_automaton import DeterministicFiniteAutomaton
```
`pyformlang` надо добавить в `requirements.txt`, если его там еще нет

DavidBykov replied:
Done

EgorOrachyov commented on tests/test_regex.py:
```diff
@@ -0,0 +1,90 @@
+from pyformlang.regular_expression import MisformedRegexError
+
+
+def test_regex_to_dfa():
+    regex = "1 1 (1 0)*"
+    dfa = regex_to_min_dfa(regex)
+    assert dfa.is_deterministic()
```
напрашивается на самом деле несколько отдельных тестов:
- детерминированный
- минимальный ? (сравнить с каким-то ожидаемым известным числом состояний)
- эквивалентный и тд

DavidBykov replied:
Проверка на минимальность и эквивалентность есть в следующем тесте

EgorOrachyov commented on tests/test_regex.py:
```diff
@@ -0,0 +1,90 @@
+    assert len(actual_dfa.states) == len(expected_dfa.states)
+
+
+@pytest.mark.parametrize(
+    "regex, expected, unexpected",
+    [
+        ("", [], [[Symbol("1")], [Symbol("0")]]),
+        (
+            "(1 0) | (1 1) | (0 1) | (0 1 0 1)*",
+            [
+                [],
+                [Symbol("1"), Symbol("0")],
+                [Symbol("1"), Symbol("1")],
+                [Symbol("0"), Symbol("1")],
+                [Symbol("0"), Symbol("1"), Symbol("0"), Symbol("1")],
+                [
+                    Symbol("0"),
+                    Symbol("1"),
+                    Symbol("0"),
+                    Symbol("1"),
+                    Symbol("0"),
+                    Symbol("1"),
+                    Symbol("0"),
+                    Symbol("1"),
+                ],
+            ],
+            [[Symbol("1")], [Symbol("0")], [Symbol("1"), Symbol("0"), Symbol("1")]],
+        ),
+        (
+            "a*",
+            [[], [Symbol("a")], [Symbol("a"), Symbol("a")]],
+            [[Symbol("b")], [Symbol("c"), Symbol("d"), Symbol("e")], [Symbol("a*")]],
+        ),
+    ],
+)
+def test_regex_to_dfa_accept(
```
Идея хорошая, но читать данную конструкцию с множеством вложенных списков затруднительно.
Предлагаю обойтись без этой аннотации (как и в других местах - там ведь тоже нет) и переделать в обычном стиле

DavidBykov replied:
Done

EgorOrachyov commented on tests/test_regex.py:
```diff
@@ -0,0 +1,90 @@
+    unexpected: List[Iterable[Symbol]],
+):
+    dfa = regex_to_min_dfa(regex)
+    for expected_word in expected:
+        assert dfa.accepts(expected_word)
+    for unexpected_word in unexpected:
+        assert not dfa.accepts(unexpected_word)
```
Немного не понятно, зачем смотреть на то, что `unexpected`, тк ожидаем только 1 результат. Предлагаю только 1 ассерт оставить

DavidBykov replied:
Я убрал, но вообще не совсем понятно, почему не нужно проверять unexpected, ведь нужно проверить, что он не пропускает слова, которые не нужны.

EgorOrachyov replied:
Теперь понятно, какая идея была с unexpected.

EgorOrachyov commented on tests/test_regex.py:
```diff
@@ -0,0 +1,90 @@
+
+
+def test_regex_to_dfa():
+    regex = "1 1 (1 0)*"
```
Достаточно простое выражение, хотелось бы еще вариантов (не обязательно с большим алфавитом), но со звездой Клини и альтернативой

DavidBykov replied:
Done

DavidBykov committed:
- Added "pyformlang" to requirements
- Added parameters to test_regex_to_dfa_accept()
- Delete unexpected words check

DavidBykov committed:
- Added more complex regex
